File: los-platform/backend/decision-engine/src/controllers/decision.controller.ts
import {
  Controller,
  Post,
  Get,
  Body,
  Param,
  Query,
  UseGuards,
  Req,
  HttpCode,
  HttpStatus,
} from '@nestjs/common';
import { ApiTags, ApiOperation, ApiResponse, ApiBearerAuth, ApiProperty } from '@nestjs/swagger';
import { DecisionEngineService } from '../services/decision-engine.service';
import {
  TriggerDecisionDto,
  ManualDecisionDto,
  DecisionResponseDto,
  OverrideRequestDto,
  OverrideApproveDto,
  OverrideRequestResponseDto,
} from '../dto';
import { RolesGuard, Roles, AuthenticatedRequest, SkipAuth } from '@los/common';

@ApiTags('Decision Engine')
@ApiBearerAuth()
@Controller('decisions')
@UseGuards(RolesGuard)
export class DecisionController {
  constructor(private readonly decisionService: DecisionEngineService) {}

  @Post('trigger')
  @Roles('CREDIT_ANALYST', 'BRANCH_MANAGER', 'ZONAL_CREDIT_HEAD', 'SYSTEM')
  @HttpCode(HttpStatus.OK)
  @ApiOperation({ summary: 'Trigger credit decision for application' })
  @ApiResponse({ status: 200, description: 'Decision result', type: DecisionResponseDto })
  async triggerDecision(
    @Body() dto: TriggerDecisionDto,
    @Req() req: AuthenticatedRequest,
  ): Promise<DecisionResponseDto> {
    return this.decisionService.triggerDecision(dto);
  }

  @Get(':applicationId')
  @Roles('CREDIT_ANALYST', 'BRANCH_MANAGER', 'ZONAL_CREDIT_HEAD', 'LOAN_OFFICER', 'COMPLIANCE_OFFICER')
  @ApiOperation({ summary: 'Get decision for application' })
  @ApiResponse({ status: 200, description: 'Decision result', type: DecisionResponseDto })
  async getDecision(
    @Param('applicationId') applicationId: string,
    @Req() req: AuthenticatedRequest,
  ): Promise<DecisionResponseDto> {
    return this.decisionService.getDecision(applicationId);
  }

  @Post('override')
  @Roles('BRANCH_MANAGER', 'ZONAL_CREDIT_HEAD')
  @HttpCode(HttpStatus.OK)
  @ApiOperation({ summary: 'Manual override of decision (direct — use maker-checker flow for production)' })
  @ApiResponse({ status: 200, description: 'Updated decision', type: DecisionResponseDto })
  async manualOverride(
    @Body() dto: ManualDecisionDto,
    @Req() req: AuthenticatedRequest,
  ): Promise<DecisionResponseDto> {
    return this.decisionService.manualOverride(dto, req.user.id);
  }

  @Post('override/request')
  @Roles('BRANCH_MANAGER', 'ZONAL_CREDIT_HEAD', 'CREDIT_ANALYST')
  @HttpCode(HttpStatus.CREATED)
  @ApiOperation({ summary: 'Request manual override (maker — creates OVERRIDE_PENDING status)' })
  @ApiResponse({ status: 201, description: 'Override request created', type: OverrideRequestResponseDto })
  async requestOverride(
    @Body() dto: OverrideRequestDto,
    @Req() req: AuthenticatedRequest,
  ): Promise<OverrideRequestResponseDto> {
    return this.decisionService.requestOverride(dto, req.user.id);
  }

  @Post('override/approve')
  @Roles('ZONAL_CREDIT_HEAD')
  @HttpCode(HttpStatus.OK)
  @ApiOperation({ summary: 'Approve or reject override request (checker — completes override)' })
  @ApiResponse({ status: 200, description: 'Override request processed', type: OverrideRequestResponseDto })
  async approveOverride(
    @Body() dto: OverrideApproveDto,
    @Req() req: AuthenticatedRequest,
  ): Promise<OverrideRequestResponseDto> {
    return this.decisionService.approveOverride(dto, req.user.id);
  }

  @Get('override/:requestId')
  @Roles('BRANCH_MANAGER', 'ZONAL_CREDIT_HEAD', 'CREDIT_ANALYST', 'COMPLIANCE_OFFICER')
  @ApiOperation({ summary: 'Get override request details' })
  @ApiResponse({ status: 200, description: 'Override request details', type: OverrideRequestResponseDto })
  async getOverrideRequest(@Param('requestId') requestId: string): Promise<OverrideRequestResponseDto> {
    return this.decisionService.getOverrideRequest(requestId);
  }

  @Get('override/pending')
  @Roles('ZONAL_CREDIT_HEAD')
  @ApiOperation({ summary: 'List all pending override requests for the approver' })
  @ApiResponse({ status: 200, description: 'List of pending override requests', type: [OverrideRequestResponseDto] })
  async listPendingOverrides(): Promise<OverrideRequestResponseDto[]> {
    return this.decisionService.listPendingOverrides();
  }
}

@ApiTags('Health')
@Controller('health')
export class HealthController {
  @SkipAuth()
  @Get()
  @ApiOperation({ summary: 'Health check' })
  @ApiResponse({ status: 200, description: 'Service is healthy' })
  check() {
    return {
      status: 'ok',
      timestamp: new Date().toISOString(),
      service: 'decision-engine',
    };
  }
}

File: los-platform/backend-java/src/main/java/com/los/decision/controller/DecisionController.java
package com.los.decision.controller;

import com.los.decision.dto.*;
import com.los.decision.entity.DecisionHistory;
import com.los.decision.service.DecisionEngineService;
import com.los.common.dto.ApiResponse;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.security.SecurityRequirement;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@Slf4j
@RestController
@RequestMapping("/api/decisions")
@RequiredArgsConstructor
@Tag(name = "Decision Engine", description = "Decision Engine APIs for credit decisions")
@SecurityRequirement(name = "bearerAuth")
public class DecisionController {

    private final DecisionEngineService decisionEngineService;

    @PostMapping("/trigger")
    @Operation(summary = "Trigger credit decision for application", description = "Run decision engine to determine loan approval/rejection")
    public ResponseEntity<ApiResponse<DecisionResponseDto>> triggerDecision(
            @Valid @RequestBody TriggerDecisionDto dto) {
        log.info("POST /api/decisions/trigger - Application: {}", dto.getApplicationId());

        DecisionResponseDto response = decisionEngineService.triggerDecision(dto);
        return ResponseEntity
                .status(HttpStatus.OK)
                .body(ApiResponse.success(response, "Decision triggered successfully"));
    }

    @GetMapping("/{applicationId}")
    @Operation(summary = "Get decision for application", description = "Retrieve decision details for specific application")
    public ResponseEntity<ApiResponse<DecisionResponseDto>> getDecision(
            @PathVariable String applicationId) {
        log.info("GET /api/decisions/{} - Fetching decision", applicationId);

        DecisionResponseDto response = decisionEngineService.getDecision(applicationId);
        return ResponseEntity
                .status(HttpStatus.OK)
                .body(ApiResponse.success(response, "Decision retrieved successfully"));
    }

    @PostMapping("/override")
    @Operation(summary = "Manual override of decision", description = "Override decision with manual approval/rejection")
    public ResponseEntity<ApiResponse<DecisionResponseDto>> manualOverride(
            @Valid @RequestBody ManualDecisionDto dto,
            @RequestHeader(value = "X-User-Id", required = false) String headerUserId) {
        log.info("POST /api/decisions/override - Application: {}", dto.getApplicationId());

        String userId = (headerUserId != null) ? headerUserId : com.los.common.security.SecurityConstants.SYSTEM_USER;
        DecisionResponseDto response = decisionEngineService.manualOverride(dto, userId);
        return ResponseEntity
                .status(HttpStatus.OK)
                .body(ApiResponse.success(response, "Decision overridden successfully"));
    }

    @PostMapping("/override/request")
    @Operation(summary = "Request manual override", description = "Create override request (maker in maker-checker flow)")
    public ResponseEntity<ApiResponse<DecisionResponseDto>> requestOverride(
            @Valid @RequestBody OverrideRequestDto dto,
            @RequestHeader(value = "X-User-Id", required = false) String headerUserId) {
        log.info("POST /api/decisions/override/request - Application: {}", dto.getApplicationId());

        String userId = (headerUserId != null) ? headerUserId : com.los.common.security.SecurityConstants.SYSTEM_USER;
        DecisionResponseDto response = decisionEngineService.requestOverride(dto, userId);
        return ResponseEntity
                .status(HttpStatus.CREATED)
                .body(ApiResponse.success(response, "Override request created successfully"));
    }

    @GetMapping("/rules")
    @Operation(summary = "List all active decision rules", description = "Get list of all active rules for admin")
    public ResponseEntity<ApiResponse<List<RuleListDto>>> listRules() {
        log.info("GET /api/decisions/rules - Listing all rules");

        List<RuleListDto> rules = decisionEngineService.listRules();
        return ResponseEntity
                .status(HttpStatus.OK)
                .body(ApiResponse.success(rules, "Rules retrieved successfully"));
    }

    @GetMapping("/status/{status}")
    @Operation(summary = "Get decisions by status", description = "Retrieve all decisions with specific status")
    public ResponseEntity<ApiResponse<List<DecisionResponseDto>>> getDecisionsByStatus(
            @PathVariable String status) {
        log.info("GET /api/decisions/status/{} - Fetching decisions", status);

        List<DecisionResponseDto> decisions = decisionEngineService.getDecisionsByStatus(status);
        return ResponseEntity
                .status(HttpStatus.OK)
                .body(ApiResponse.success(decisions, "Decisions retrieved successfully"));
    }

    @GetMapping("/{applicationId}/history")
    @Operation(summary = "Get decision history", description = "Retrieve audit trail of decision changes")
    public ResponseEntity<ApiResponse<List<DecisionHistory>>> getDecisionHistory(
            @PathVariable String applicationId) {
        log.info("GET /api/decisions/{}/history - Fetching history", applicationId);

        List<DecisionHistory> history = decisionEngineService.getDecisionHistory(applicationId);
        return ResponseEntity
                .status(HttpStatus.OK)
                .body(ApiResponse.success(history, "Decision history retrieved successfully"));
    }
}
